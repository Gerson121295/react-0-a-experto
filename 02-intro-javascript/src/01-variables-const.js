
//Este archivo.js almacena lo realizado en main.js para ejecutarlo pegar a main.js y levantar la app

/*** Variables  y constantes ***/
//var para definir variables y constantes ya no se utilizan
const nombre = 'Pepe'; 
let apellido = 'Rdas'

let valorDado = 5;
valorDado = 6;

console.log("-----------Imprime dato dentro del if------------- ")
if ( true ){
  const nombre = "Sara";
  let valorDado = 3;
  console.log(nombre + " " + valorDado);
}

console.log("------------- Imprime dato de variables----------- ")
console.log(nombre + '  ' + apellido + ' ' + valorDado);
console.log("------------- Imprime dato de variables usando Template String----------- ")
console.log(`nombre ${nombre}  apellido ${apellido} valor Dado ${valorDado}`);


//function getSaludo(nombre){  //Funcion norma
  const getSaludo = () => {  //usando arrow function 
  return 'hola '+ nombre;
}

console.log(`Este es my name: ${getSaludo(nombre)}`); //Llama funcion dentro de Template String
