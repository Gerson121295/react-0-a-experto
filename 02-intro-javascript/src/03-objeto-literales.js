

//Objetos Literales

const persona = {
    nombre: 'Toni',
    apellido: 'Stark',
    edad: 45,
  };
  
  console.log(persona.nombre);
  console.log({persona:persona}); //se crea un objeto persona y apunta al valor objeto persona, otra forma de escribirlo cuando se llama igual es solo agregar 1 vez
  console.log({persona}); //se crea un objeto persona y apunta al valor objeto persona, como se llama igual se escribe 1 vez
  //console.table({persona}); //table para una mejor vista del objeto persona
  
  //Los objetos pueden tener dentro otros objetos
  const people = {
    nombre: 'Toni',
    apellido: 'Stark',
    edad: 45,
    //Objeto direccion dentro de people
    direccion:{
      ciudad: 'New york',
      zip: 55321221,
      lat: 14.3232,
      lng: 34.9233321,
    }
  };
  
  console.log(people);
  
  //Clonar el objeto people
  console.log("----------Clon de people -----------");
  const people2 = people; //no se copia el valor de persona, copia la referencia espacio en memoria donde se guarda people
        people2.nombre = 'Peter';
  console.log(people2);
  
  //Para clonar un objeto lo recomendable es crear un objeto y agregar las propiedades
  const people3 = {nombre: 'Pablo', apellido: 'Lu', edad:'21'}
  
  //Que pereza es agregar todas las propiedades del objeto para crear un clon por lo que se utiliza el operador Spread para agregarlas
  console.log("------ Crear un clon de people utilizando el operador ... spread ------")
  const people4 = {...people}; //en people4 se crea un clon del objeto people con sus datos.
        people4.nombre = 'Luis'; //se modifica solo el nombre del objeto people4
        console.log(people4);

        