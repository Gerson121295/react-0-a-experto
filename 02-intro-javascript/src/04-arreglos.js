


//Arreglos [] - en JS

//Forma 1 de crear arreglo - No recomendable porque tiene tama침o
const arreglo = new Array(100); //el arreglo tengra un tama침o de 100
arreglo.push(1); //agrega una cantida al arreglo ahora sera de 101
console.log(arreglo);

//Forma 2 de crear arreglo 
console.log("--------Crear arreglo sin tama침o definido --");
const arreglo1 = [];
arreglo1.push(1); //insertar al arreglo el 1  - no es recomendable usar push para agregar datos al arreglo
arreglo1.push(2); 
console.log(arreglo1);

//Forma 3 crear arreglo
let arreglo2 = [1,2,3,4]
console.log(arreglo2);

console.log("4 - Crear arreglo y agregar datos de arreglo2 mas asigna nuevos datos");
let arreglo3 = [...arreglo2, 5]; //crea clon(con spread) del arreglo2 en arreglo3 y le agrega un nuevo dato 5
console.log(arreglo3);

console.log("*-------------------*");
//"Utiliando el map (crear un nuevo arreglo el cual se modifica pero no modifica el arreglo que se le envia)");
const arreglo4 = arreglo3.map(function(number){ //callback esta funcion recibe un number y se ejecutar치 por cada elemento del arreglo3
    return number + 2; //cada elemento del arreglo3 se sumara 2
});
console.log("arreglo3 no modificado: " + arreglo3);
console.log("arreglo4 map modificado: " + arreglo4);


