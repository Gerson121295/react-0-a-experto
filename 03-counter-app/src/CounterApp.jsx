
import PropTypes from 'prop-types'
import { useState } from 'react'

export const CounterApp = ({value}) => { //recibe props value del padre main.js
    
    console.log('render'); //cuando cambia el estado el componente se vuelve a ejecutar por lo que se imprime en consola "render"

    //UseState- valor counter, con setCounter se cambia el estado o valor de counter , useState(0) define que counter inicia con valor 0
    const [counter, setCounter] = useState(value) //useState(0);  //useState(value) counter inicia el conteo con el valor value(100) que recibe del padre main.jsx

    //Definir funcion para aumentar en 1 al presionar el boton
    const aumentar=()=>{
        setCounter(counter +1)
    }

    //Funcion al dar clic en boton despliega todo las propiedades de event para manejarlo
    function handledAdd(){ //const handleAdd = (event) => {   //console.log(event)
        //console.log(event)
        //setCounter(counter +1)
        setCounter((c) => c+1) //forma 2 de aumentar: obtiene el callback c y le sigue aumentando 1
    }

    //Funcion para decrementar
    const handleSubtract = () => setCounter(counter - 1);

    //Funcion para reset
    const handleReset = () => setCounter(value) //(counter==value); 

    return(
        <>
            <h1>CounterApp</h1>
            {/* <h2>{value}</h2> */}
            <h2>{counter}</h2>
            <button 
                className='button'
                //onClick={() => setCounter(counter + 1)} //cada vez que precione el boton llama la funcion onClic y setCounter(modifica counter) al ejecutar "counter+1" se aumenta en 1
                //onClick={() => aumentar()} //forma 2 de aumentar: definir una funcion afuera y llamarla
                //onClick={aumentar} //forma 3 de aumentar:
                 onClick={handledAdd} //aumenta el valor
            >
                +1
            </button>
            <button
                className='button'
                onClick={handleSubtract}
            >
                -1
            </button>

            <button
                aria-label='btn-reset' //es el id para las pruebas, otra forma es: id='btn-reset' 
                className='button'
                onClick={handleReset}
            >
                Reset
            </button>
        </>
    )
}

CounterApp.propTypes = {
    value: PropTypes.number.isRequired,
}

