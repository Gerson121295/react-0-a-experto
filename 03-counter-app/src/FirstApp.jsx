import PropTypes from 'prop-types'


//Recomendable agregar variables o funciones que no tengan nada que ver con la funcionalidad a fuera de la funcion 
//const newMessage = 'Gerson';
//const newMessage = [1,2,3,4,5,6,7];

/* const newMessage = {
     message: 'Hola',
    title: 'Gep'
} 
*/

function suma(a,b){
    return a + b;
}

export const FirstApp = ({title='No hay title', subTitle='No hay subtitle', name='Gema'}) =>{ //recibe props del padre main.jsx: {title} esta desestructurado de props

    return(
        <>
            {/* <h1>GErson Ep</h1> */}
            {/* <p>{newMessage.message}</p> */} {/* impresion de variables en el HTML */}
            {/* <code>{JSON.stringify(newMessage)}</code>  */}{/* impresion de objeto js */}
            {/* {<p>La suma es: {suma(5,5)}</p>} */}

            {/* Recibe datos mediante props */}
            {/* Title es recibida como props */}
            <h1 data-testid="test-title">{title}</h1> 
            <p>{subTitle}</p> {/* <p>{subTitle} {suma(5,5)}</p>  */}
            <p>{subTitle}</p> 
            <h2>{name}</h2>
        </>
    )
}

/*
export function FirstComponent(){
    return(
        <h3>First component number 2</h3>
    )
}
*/

//Validar las props que se recibe del padre cumpla con el tipo de dato requerido
FirstApp.propTypes ={
    //name: PropTypes.string,
    title: PropTypes.string.isRequired,
    subTitle: PropTypes.string,
    name: PropTypes.string,
    
}


//Definir los props por defecto en caso de que no viene del padre pero si viene tomará la del padre
//Para futuras versiones ya no se utilizará .defaultProps = {  
//para establecer props por defecto se realizará en la funcion donde se define la props a recibir del padre:
//export const FirstApp = ({ title = 'No hay título', subTitle = 'No hay subtítulo' }) => {...}
/*
FirstApp.defaultProps = {
    name: 'Gerson Ep',
    subTitle: 'No hay subtitulo',
    //title: "No hay titulo"
}
*/
export default FirstApp;