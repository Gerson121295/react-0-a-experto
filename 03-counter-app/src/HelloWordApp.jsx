


//Diferentes formas de crear funciones y exportarlas a main.jsx
export function App(){
    return (<h1>hola Mundo!!!</h1>);
}



export const HelloWordApp = () =>{
    return(
        <h2>224</h2>
    )
} 
   