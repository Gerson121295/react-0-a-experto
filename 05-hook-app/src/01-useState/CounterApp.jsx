import { useState } from "react"

export const CounterApp = () => {

    //useState: counter es el valor, con setCounter se encarga de modificar counter y el useState(10) inicializa a counter en 10
    //const [counter, setCounter] = useState(10);
    //const incrementar = () => { setCounter(counter + 1) }


    //multicounter
     //const [counter, setCounter] = useState({ counter1: 10, counter2: 20, counter3: 30})
     //const [{counter1, counter2, counter3}, setCounter] = useState({ counter1: 10, counter2: 20, counter3: 30});
     
     const [state, setCounter] = useState({ counter1: 10, counter2: 20, counter3: 30})
     const {counter1, counter2, counter3} = state; //desestructura state y se guarda la data en counter1, counter2 y counter3
    
     console.log(counter1);
    return(
        <>
        
        <h1>Counter: {counter1} </h1>
        <h1>Counter: {counter2} </h1>
        <h1>Counter: {counter3} </h1>
        <hr />
        

        <button
        className="btn"
        //onClick={() => setCounter(counter + 1 ) }
        //onClick={incrementar} // onClick={(counter) => incrementar(counter)} //cuando el argumento que se recibe en el callback es el mismo parametro, que se va a enviar en la funcion se reduce agregando solo el nombre de la funcion.
          
        //onClic para multicounter
        //onClick={() => setCounter({counter1: counter1+1, counter2:counter2, counter3})} //conter3- Cuando una variable es igual a la otra se puede escribir una vez
        //onClick={() => setCounter(prevState => ({ ...prevState, counter1: prevState.counter1 + 1 })) }
        
        //multiconter utilizando state y desestructuracion
        onClick={() => setCounter({ ...state, counter1: counter1 + 1 })} //...state pasa los valores anteriores y a counter1: counter1+1 se le suma 1

        >

            +1
        </button>

        

        </>
        
    )
    
}