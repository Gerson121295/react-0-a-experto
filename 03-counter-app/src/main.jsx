import React from 'react'
import ReactDOM from 'react-dom/client';
import './styles.css'
import { CounterApp } from './CounterApp';

/* function App(){ //se puede definir funciones aqui.
return (<h1>hola Mundo</h1>);
} */

//Manejando el DOM en React: root definido en index.html
ReactDOM.createRoot(document.getElementById('root')).render(

    <React.StrictMode>
        {/* Renderiza la funcion App - aqui se llama a renderizar otro components */}
        

{/*     <FirstApp
        //Del component padre main.jsx Envia prop title y subtitle al hijo FirstApp
        title={"Hola soy Vegeta"}
        subTitle={"La suma es:"}
        name={"Gerson Ep"}
        numero={123} //enviar tipo numero
        />
 */}
       

    <CounterApp 
         value={0}
    />

    </React.StrictMode>
);


