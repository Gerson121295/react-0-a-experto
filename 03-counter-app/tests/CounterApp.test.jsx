import { fireEvent, render, screen } from "@testing-library/react";
import { CounterApp } from "../src/CounterApp"

describe(' Pruebas a CounterApp', () => {

    //Variables globales
    const initialValue = 10;

    test('Pruebas en el <CounterApp /> ', () => {

        const { container} = render(<CounterApp value={initialValue} /> );
        expect(container).toMatchSnapshot();

    });

    
    test('Debe de mostrar el valor inicial de 100 <CounterApp value={100}/>', () => {
        
        render(<CounterApp value = {100} /> );
        //expect(screen.getAllByText(100)).toBeTruthy();//valida que value sea 100 true pero puede haber mas value con valor de 100
       expect(screen.getByText(100)).toBeTruthy();//valida que vsolo halla un value y sea 100 true
        
       //expect(screen.getByRole('heading',{level:2}).innerHTML ).toContain('100'); //valida que exista un h2 y que tenga valor a 100
    });

    test('Debe de incrementar con el boton +1', () => {
        render(<CounterApp value={initialValue} /> ); //valor inicial que se envia es 10 como props
        
        //al dar clic en el boton '+1' aumente la cantidad del contador
        fireEvent.click(screen.getByText('+1')); //para saber que boton es: busca por texto '+1' y da clic
        expect(screen.getByText('11')).toBeTruthy(); //luego de dar clic en '+1' aparecera un texto con valor de 11
    });

    test('debe de decrementar con el boton -1', () => {
        render(<CounterApp value={initialValue} /> ); //valor inicial que se envia es 10 como props
        
        //al dar clic en el boton '-1' decrementa la cantidad del contador
        fireEvent.click(screen.getByText('-1')); //para saber que boton es: busca por texto '-1' y da 1 clic
        screen.debug(); //para ver el estado del dom
        expect(screen.getByText('9')).toBeTruthy(); //luego de dar clic en '-1' aparecera un texto con valor de 9
    });

    test('debe de funcionar el boton reset', () => {
       // render(<CounterApp value={initialValue} /> ); //valor inicial que se envia es 10 como props
       render(<CounterApp value={350} /> ); //valor inicial que se envia es 10 como props
        
         //al dar clic en el boton '+1' aumente la cantidad del contador
         fireEvent.click(screen.getByText('+1')); //busca por texto '+1' da un clic queda en 11
         fireEvent.click(screen.getByText('+1')); //busca por texto '+1' da un clic queda en 12
         fireEvent.click(screen.getByText('+1')); //busca por texto '+1' da un clic queda en 13
      
         
        //al dar clic en el Reset resetea la cantidad del contador al valor inicial 10
        //fireEvent.click(screen.getByText('Reset')); //para saber que boton es: busca por texto 'Reset' luego da clic y resetea el valor
        //screen.debug(); //para ver el estado del dom
        
        //Otra forma de buscar el boton y darle clic usando: aria-label='btn-reset' su key es btn-reset definido en etiqueta Button html de CounterApp.jsx
        fireEvent.click(screen.getByRole('button', {name:'btn-reset'})); 

        expect(screen.getByText(350)).toBeTruthy(); //luego de dar clic en Reset se valida si existe un valor con el valor inicial si existe entonces Ok.
        //expect(screen.getByText(initialValue)).toBeTruthy(); //luego de dar clic en Reset se valida si existe un valor con el valor inicial que es 10. si es true esta ok
    
    });



});


