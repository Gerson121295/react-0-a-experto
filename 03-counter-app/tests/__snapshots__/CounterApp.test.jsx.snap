// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Pruebas a CounterApp Pruebas en el <CounterApp />  1`] = `
<div>
  <h1>
    CounterApp
  </h1>
  <h2>
    10
  </h2>
  <button
    class="button"
  >
    +1
  </button>
  <button
    class="button"
  >
    -1
  </button>
  <button
    aria-label="btn-reset"
    class="button"
  >
    Reset
  </button>
</div>
`;
